### PR TITLE
[P4-1976] Support event includes in all known eventables

### PR DESCRIPTION
**Pull request by willfish**
### Jira link

P4-1976

### What?

I have added/removed/altered:

- [x] Added has_many :events serializer include support to all known eventables

### Why?

This is for the frontend to be able to get events to be displayed on a timeline.

We were initially musing the idea of adding an endpoint for indexing events which you could filter for multiple eventables. This was going to require extra plumbing for the frontend and us and didn't really follow any patterns around polymorphic filters that the json:api spec defined. We can comfortably rely on includes for existing eventables and have the frontend stitch them together/reuse them in their workflows.

### Have you? (optional)

- [x] I'm not going to add these includes to the swagger documentation as this is meant to be exclusively used by the frontend.

### CODE REVIEW

**Review by cwrw — COMMENTED** [review]
including events could have performance implications on endpoints, there is some work being done on eager/lazy loading associations, I think we should add events to this as well. Also could you update the swagger docs to include the new includes?

File: app/models/person.rb [hunk 1]
```diff
@@ -18,6 +18,8 @@ class Person < VersionedModel
 
   has_many :profiles, dependent: :destroy
   has_many :moves, through: :profiles
+  has_many :generic_events, as: :eventable, dependent: :destroy # NB: polymorphic association
```
Comment (cwrw) [inline app/models/person.rb:21]: can you add a test for this to maintain the pattern we have for associations on models?

Comment (willfish) [reply app/models/person.rb:21]: Sure.

**Comment by willfish** [comment]
@cwrw See my description re swagger docs

File: app/serializers/journey_serializer.rb [hunk 2]
```diff
@@ -11,6 +11,10 @@ class JourneySerializer
   has_one :from_location, serializer: LocationSerializer
   has_one :to_location, serializer: LocationSerializer
 
+  has_many :events, serializer: GenericEventSerializer do |object|
+    object.generic_events.applied_order
```
Comment (cwrw) [inline app/serializers/journey_serializer.rb:15]: will using `applied_order` scope cause n+1 queries on the eventables?

File: app/serializers/v2/move_serializer.rb [hunk 3]
```diff
@@ -29,6 +29,9 @@ class MoveSerializer
     has_one :supplier, serializer: SupplierSerializer
 
     has_many :court_hearings, serializer: CourtHearingSerializer
+    has_many :events, serializer: GenericEventSerializer do |object|
```
Comment (cwrw) [inline app/serializers/v2/move_serializer.rb:32]: do we not want to add this to v1 as well, as its not a breaking change?

Comment (willfish) [reply app/serializers/v2/move_serializer.rb:32]: No no. V1 is dead and if people want events to be surfaced they can move onto v2

File: spec/serializers/person_escort_record_serializer_spec.rb [hunk 4]
```diff
@@ -137,4 +137,41 @@
...
+
+      let(:expected_event_relationships) do
+        [
+          { id: third_event.id, type: 'events' },
+          { id: second_event.id, type: 'events' },
+          { id: first_event.id, type: 'events' },
+        ]
+      end
+
+      it 'contains event relationships' do
```
Comment (cwrw) [inline spec/serializers/person_escort_record_serializer_spec.rb:158]: could you add something in the test description to indicate that order matters here? (Same everywhere)

Comment (willfish) [reply spec/serializers/person_escort_record_serializer_spec.rb:158]: Sure

**Comment by cwrw** [comment]
I think its worth documenting the relationships at least as suppliers might inquire to what they are when they see those relationships in our resources when querying the endpoints. We don't need to add the those to the include list though

File: app/serializers/journey_serializer.rb (hunk 2 above)
Comment (willfish) [reply app/serializers/journey_serializer.rb:15]: Possibly. I'll talk to Martyn to avoid this situation when indexing moves.

Comment (willfish) [reply app/serializers/journey_serializer.rb:15]: Looking at the output from bullet it doesn't complain about n+1s here 

**when showing an individual move:**

```
GET /api/moves/d17c348f-ae1c-4385-9899-0b711fe17e19?include=events
AVOID eager loading detected
  Move => [:from_location, :to_location, :profile]
  Remove from your query: .includes([:from_location, :to_location, :profile])
Call stack

2020-10-21 16:38:33[WARN] user: william
GET /api/moves/d17c348f-ae1c-4385-9899-0b711fe17e19?include=events
AVOID eager loading detected
  Profile => [:person, :person_escort_record]
  Remove from your query: .includes([:person, :person_escort_record])
Call stack

2020-10-21 16:38:33[WARN] user: william
GET /api/moves/d17c348f-ae1c-4385-9899-0b711fe17e19?include=events
AVOID eager loading detected
  Person => [:gender, :ethnicity]
  Remove from your query: .includes([:gender, :ethnicity])
Call stack
```

**when indexing moves:**

```
GET /api/moves?include=events
AVOID eager loading detected
  Move => [:allocation, :supplier, :court_hearings, :prison_transfer_reason, :original_move, :person, :from_location, :to_location]
  Remove from your query: .includes([:allocation, :supplier, :court_hearings, :prison_transfer_reason, :original_move, :person, :from_location, :to_location])
Call stack

2020-10-21 16:40:28[WARN] user: william
GET /api/moves?include=events
AVOID eager loading detected
  Profile => [:documents, :person_escort_record]
  Remove from your query: .includes([:documents, :person_escort_record])
Call stack

2020-10-21 16:40:28[WARN] user: william
GET /api/moves?include=events
AVOID eager loading detected
  Document => [:file_attachment]
  Remove from your query: .includes([:file_attachment])
Call stack
```

**Review by cwrw — APPROVED** [review]
Thanks for addressing the comments, looks good. There is a missing include in journeys, otherwise 👍

File: app/serializers/journey_serializer.rb [hunk 5]
```diff
@@ -11,6 +11,10 @@ class JourneySerializer
...
+  end
+
   SUPPORTED_RELATIONSHIPS = %w[
     from_location
     from_location.suppliers
```
[the remark this answers is not in the document]
Comment (willfish) [reply app/serializers/journey_serializer.rb:20]: Yes we do. Thanks

**Pull request merged**